### PR TITLE
Generate production locales using 'compare-locales'

### LINKED ISSUE #357: Lint some locales?
@flodolo, @mathjazz, and @Pike pointed me at [**compare-locales**](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/compare-locales) the other day:

- https://github.com/mozilla/send/issues/330#issuecomment-318382224
- https://github.com/mozilla/send/issues/335#issue-246066911

Not sure if it'd be super useful to add some `lint:locales` npm script, or maybe just a separate `compare-locales` script that either validates all locales, or maybe just lint the production locales so we can check that all the prod strings are error free before doing deploys or whatever.

Here was a borderline gross pile of code I threw together that will lint locales using the Python `compare-locales` tool (which we'd need to install on Circle-CI) and displays just the error messages:

```js
#!/usr/bin/env node

const { exec } = require('child_process');
const fs = require('fs');
const path = require('path');

const STATIC_PATH = path.join(__dirname, '../public');

function allLangs() {
  return fs
    .readdirSync(path.join(STATIC_PATH, 'locales'))
    .join(' ');
}

function prodLangs() {
  return require('../package.json').availableLanguages.sort().join(' ');
}

const cmd = `compare-locales l10n.toml . ${allLangs()} --data=json`;

exec(cmd, (err, stdout, stderr) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }
  const details = JSON.parse(stdout).details;
  const results = filterErrors(details);
  
  if (results.length) {
    const res = JSON.stringify(results, null, 2);
    console.info(res);
    process.exit(2);
  }
});

function filterErrors(details) {
  return Object.keys(details)
    .sort()
    .map(locale => {
      const data = details[locale]
        .filter(item => item.hasOwnProperty('error'))
        .map(item => item.error);
      return { [locale]: data };
    })
    .filter(locale => Object.values(locale).pop().length);
}
```

Currently I'm getting errors for two locales (but haven't checked _master_ branch yet to see if they've been resolved -- plus it's only for the dev-only "nn-NO" and "tr" locales):

```sh
$ ./scripts/compare-locales-cp.js
[
  {
    "nn-NO/send.ftl": [
      "Missing attribute: title at line 10, column 1 for uploadPageBrowseButton",
      "Missing attribute: title at line 30, column 1 for copyUrlFormButton",
      "Missing attribute: title at line 33, column 1 for deleteFileButton",
      "Missing attribute: title at line 35, column 1 for sendAnotherFileLink",
      "Missing attribute: title at line 44, column 1 for downloadButtonLabel",
      "Missing attribute: title at line 48, column 1 for sendYourFilesLink"
    ]
  },
  {
    "tr/send.ftl": [
      "Missing attribute: title at line 8, column 1 for uploadPageBrowseButton",
      "Missing attribute: title at line 17, column 1 for uploadingPageCancel",
      "Missing attribute: title at line 26, column 1 for copyUrlFormButton",
      "Missing attribute: title at line 29, column 1 for deleteFileButton",
      "Missing attribute: title at line 31, column 1 for sendAnotherFileLink",
      "Missing attribute: title at line 38, column 1 for downloadButtonLabel"
    ]
  }
]


$ echo $? # 2
```

### LINKED ISSUE #364: Improve production locale detection
Currently, I believe we're manually checking production-ready locales against https://pontoon.mozilla.org/projects/test-pilot-firefox-send/ (sorting by progress to group by 100% complete), then manually updating the [package.json's `availableLanguges`](https://github.com/mozilla/send/blob/81be38fbc8b173db97c3272c63d252b72ad98aff/package.json#L48-L67) array.

It may be nice to try and find a more automated way. Here's a brief conversation from the #l20n channel on possible solutions:

> \<pdehaan\> Another l20n question, while I wait for GitHub to return from the dead... for https://pontoon.mozilla.org/projects/test-pilot-firefox-send/ we have several locales that are 100% (yay!). Is there an easy way to get that information, an API or something? Or is the best way to determine our production locales to sort the translation % in descending order manually and build the locales list?
>
> \<gandalf\> pdehaan: stas just got back from PTO and he's working on an API for our l10n infra. It may have an endpoint for that data eventually. but seems like for now manual is the way
>
> \<pdehaan\> awesome, thanks
>
> \<Pike\> pdehaan: you can run compare-locales, it comes with json output via `--data=json`
> 
> \<pdehaan\> Pike: then just loop over `summary` locale keys and ignore any locales that have a `missing` or `missing_w` prop?
> 
> \<flod\> missing would be enough (missing_w = missing words)
>
> \<Pike\> pdehaan: or some more permissive criteria, but yes

---

Being the curious lad I am, I hacked this 🗑🔥 code, which seems to work (assuming you have Python [compare-locales](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/compare-locales) installed):

```js
const { exec } = require('child_process');

const { availableLanguages } = require('../package.json');

const compareLocales = 'compare-locales l10n.toml . `ls public/locales` --data=json';

exec(compareLocales, (err, stdout, stderr) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }
  const missingLocales = (current, package) => current.filter(locale => !package.includes(locale));
  const { summary } = JSON.parse(stdout);
  const locales = Object.keys(summary)
    .filter(locale => {
      const loc = summary[locale];
      const hasMissing = loc.hasOwnProperty('missing');
      const hasErrors = loc.hasOwnProperty('errors');
      return !hasMissing && !hasErrors;
    })
    .sort();
  console.log("current 100%:", JSON.stringify(locales));
  console.log("package.json:", JSON.stringify(availableLanguages.sort()));
  console.log("missing prod:", JSON.stringify(missingLocales(locales, availableLanguages).sort()));
});
```

```sh
$ node scripts/get-prod-locales.js

current 100%: ["cs","cy","de","dsb","en-US","fr","fy-NL","hsb","hu","it","ja","kab","ms","nb-NO","nn-NO","pt-BR","pt-PT","ru","sk","sl","sv-SE","zh-CN","zh-TW"]
package.json: ["cs","cy","de","dsb","en-US","fr","fy-NL","hsb","hu","it","ja","kab","ms","pt-PT","sk","sv-SE","zh-CN","zh-TW"]
missing prod: ["nb-NO","nn-NO","pt-BR","ru","sl"]
```

### PR DESCRIPTION
**Pull request by pdehaan**
You can get the latest production locales (100% translations, zero errors) with `npm run get-prod-locales` or you can overwrite the package.json `availableLanguages` array with `npm run get-prod-locales -- --write`. We could make that a bit prettier by wrapping in another npm script, but hey...

Fixes #357 
Fixes #364

### CODE REVIEW

**Comment by dannycoates** [comment]
awesome! Should we add this to the circle build?

**Comment by pdehaan** [comment]
I have no idea when/if we want this. It almost feels weird to do this "arbitrary" decision in CI. I'm still trying to understand the implications, so I'd lean more towards managing these somewhat manually, so we have a hand-crafted, organic, artisanal selection of available locales. 

I think the problem is that if one of us adds a new string, then commits it to en-US, all locales would fall to 99% and then be excluded from stage/prod until the localizers added that one missing string.

**Comment by pdehaan** [comment]
Plus, this step requires `compare-locales` Python tool, so we'd need to make sure we could add Python to Circle-CI, just to run this one script. Meh.

File: package.json [hunk 1]
```diff
@@ -46,40 +46,38 @@
   "license": "MPL-2.0",
   "repository": "mozilla/send",
   "availableLanguages": [
-    "en-US",
```
Comment (pdehaan) [inline package.json]: @dannycoates Are these available languages in a "specific" order? I think my script just blindly A-z the locales, but wasn't sure if having "en-US" at the front is important, or if it serves as a fallback locale.

Comment (dannycoates) [inline package.json]: not really, they're in the order the pontoon ui puts them. a-z is fine

**Pull request merged**